include/aspect/melt.h

```diff
@@ -0,0 +1,279 @@
+                                   const unsigned int q_point) const;
+
+        /**
+         * Compute the right-hand side for the advection system index. This is
+         * 0 for the temperature and all of the compositional fields, except for
+         * the porosity. It includes the melting rate and a term dependent
+         * on the density and velocity.
+         *
+         * This function is implemented in
+         * <code>source/simulator/assembly.cc</code>.
+         */
+        double
+        compute_melting_RHS(const internal::Assembly::Scratch::AdvectionSystem<dim>  &scratch,
+                            typename MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
+                            typename MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                            const typename Simulator<dim>::AdvectionField &advection_field,
+                            const unsigned int q_point) const;
+    };
+
+    /**
+     * A namespace for the definition of functions that implement various
+     * other terms that need to occasionally or always be assembled.
+     */
+    namespace OtherTerms
+    {
+      /**
+       * Integrate the local fluid pressure shape functions on a single cell
+       * for models with melt migration, so that they can later be used to do
+       * the pressure right-hand side compatibility modification.
+       */
+      template <int dim>
+      void
+      pressure_rhs_compatibility_modification_melt (const SimulatorAccess<dim>                      &simulator_access,
+                                                    internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                    internal::Assembly::CopyData::StokesSystem<dim> &data);
+    }
+  }
+
+
+  /**
+   * Class to contain all runtime parameters and other helper functions
+   * related to melt transport. A global instance can be retrieved with
+   * MeltHandler<dim>::get(), but keep in mind that it only exists if
+   * parameters.include_melt_transport is true.
+   */
+  template <int dim>
+  class MeltHandler: public SimulatorAccess<dim>
+  {
+    public:
+      MeltHandler(ParameterHandler &prm);
+
+      static void declare_parameters (ParameterHandler &prm);
+      void parse_parameters (ParameterHandler &prm);
+
+      void edit_finite_element_variables(const Parameters<dim> &parameters, std::vector<VariableDeclaration<dim> > &variables);
+      void initialize_simulator (const Simulator<dim> &simulator_object);
+
+      /**
+       * Compute fluid velocity and solid pressure in this ghosted solution vector.
+       * @param solution
+       */
+      void compute_melt_variables(LinearAlgebra::BlockVector &solution);
+
+      /**
+       * Return whether this object refers to the porosity field.
+       */
+      bool is_porosity (const typename Simulator<dim>::AdvectionField &advection_field) const;
```

Comment:
Elegant workaround to not add more and more `is_...` functions to `AdvectionField` :+1: